src/Pos/Delegation/Helpers.hs

```diff
@@ -21,3 +34,60 @@ dlgVerifyPayload epoch (getDlgPayload -> proxySKs) =
     throwError "Block contains psk(s) that have non-matching epoch index"
   where
     notMatchingEpochs = filter ((/= epoch) . pskOmega) proxySKs
+
+-- | Checks if given PSK revokes delegation (issuer = delegate).
+isRevokePsk :: ProxySKHeavy -> Bool
+isRevokePsk ProxySecretKey{..} = pskIssuerPk == pskDelegatePk
+
+-- | Applies block certificates to 'ProxySKHeavyMap'.
+dlgMemPoolApplyBlock :: MainBlock ssc -> DlgMemPool -> DlgMemPool
+dlgMemPoolApplyBlock block m = flip execState m $ do
+    let (toDelete,toReplace) =
+            partition isRevokePsk (getDlgPayload $ block ^. mainBlockDlgPayload)
+    for_ toDelete $ \psk -> identity %= HM.delete (pskIssuerPk psk)
+    for_ toReplace $ \psk -> identity %= HM.insert (pskIssuerPk psk) psk
+
+-- | Checks if addition of the PSK to the map will lead to cycles. The
+-- initial map may or may not contain this PSK. Returns nothing if
+-- it's good, first-already-visited public key otherwise.
+dlgMemPoolDetectCycle
+    :: forall m . (Monad m)
+    => (PublicKey -> m (Maybe ProxySKHeavy)) -- ^ Resolving function
+    -> ProxySKHeavy                          -- ^ PSK to check against
+    -> m (Maybe PublicKey)
+dlgMemPoolDetectCycle resolve toAdd =
+    evalStateT (trav (pskDelegatePk toAdd)) (HS.singleton $ pskIssuerPk toAdd)
+  where
+    trav :: PublicKey -> StateT (HashSet PublicKey) m (Maybe PublicKey)
+    trav cur = ifM (uses identity $ HS.member cur) (pure $ Just cur) $ do
+        next <- lift $ resolve cur
+        identity %= HS.insert cur
+        let stop = pure Nothing
+        maybe stop (\psk -> bool stop (trav $ pskDelegatePk psk) $ isRevokePsk psk) next
+
+-- | Given an 'DlgMemPool', issuer, delegate and his psk, checks if
+-- delegate is allowed to issue the block. This does not check that
+-- delegate didn't issue a psk to somebody else.
+dlgReachesIssuance
```

Comment:
@volhovM 
hm, suppose there three nodeы: A, B, C
A -> B (A delegated stake to B)
B -> C
1. can B publish blocks at all?
2. can B publish block instead of A? or only C can do it?